godot/gdfreeorion.cpp

```diff
@@ -0,0 +1,120 @@
+#include <sstream>
+
+#include "../util/Directories.h"
+#include "../util/OptionsDB.h"
+#include "../util/Version.h"
+#include "../client/ClientNetworking.h"
+
+#include "GodotClientApp.h"
+#include "OptionsDB.h"
+
+using namespace godot;
+
+std::atomic_bool quit(false);
+
+void GDFreeOrion::do_the_ping(GDFreeOrion* n) {
+    while(!quit) {
+        if (auto msg = n->app->Networking().GetMessage()) {
+            std::ostringstream stream;
+            stream << msg->Type();
+            n->emit_signal("ping", String(stream.str().c_str()));
+        } else {
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+    }
+}
+
+void GDFreeOrion::_register_methods() {
+    // register_method("_process", &GDCppTest::_process);
+    register_method("_exit_tree", &GDFreeOrion::_exit_tree);
+    register_signal<GDFreeOrion>("ping", "message", GODOT_VARIANT_TYPE_STRING);
+    register_property<GDFreeOrion, godot::OptionsDB*>("optionsDB",
+        &GDFreeOrion::set_options,
+        &GDFreeOrion::get_options,
+        nullptr);
+    register_property<GDFreeOrion, GodotNetworking*>("networking",
+        &GDFreeOrion::set_networking,
+        &GDFreeOrion::get_networking,
+        nullptr);
+}
+
+GDFreeOrion::GDFreeOrion() {
+}
+
+GDFreeOrion::~GDFreeOrion() {
+    // add your cleanup here
+}
+
+void GDFreeOrion::_init() {
+    InitDirs("");
+
+    GetOptionsDB().SetFromFile(GetConfigPath(), FreeOrionVersionString());
+    GetOptionsDB().SetFromFile(GetPersistentConfigPath());
+
+    CompleteXDGMigration();
+
+    // Handle the case where the resource.path does not exist anymore
+    // gracefully by resetting it to the standard path into the
+    // application bundle.  This may happen if a previous installed
+    // version of FreeOrion was residing in a different directory.
+    if (!boost::filesystem::exists(GetResourceDir()) ||
+        !boost::filesystem::exists(GetResourceDir() / "credits.xml") ||
+        !boost::filesystem::exists(GetResourceDir() / "data" / "art" / "misc" / "missing.png"))
+    {
+        DebugLogger() << "Resources directory from config.xml missing or does not contain expected files. Resetting to default.";
+
+        GetOptionsDB().Set<std::string>("resource.path", "");
+
+        // double-check that resetting actually fixed things...
+        if (!boost::filesystem::exists(GetResourceDir()) ||
+            !boost::filesystem::exists(GetResourceDir() / "credits.xml") ||
+            !boost::filesystem::exists(GetResourceDir() / "data" / "art" / "misc" / "missing.png"))
+        {
+            DebugLogger() << "Default Resources directory missing or does not contain expected files. Cannot start game.";
+            throw std::runtime_error("Unable to load game resources at default location: " +
+                                     PathToString(GetResourceDir()) + " : Install may be broken.");
+        }
+    }
+
+    // initialize any variables here
+    app = std::make_unique<GodotClientApp>();
```

Comment:
That line needs to be moved up, right after the call to `CompleteXDGMigration` (line 58). Otherwise the logging system does not get initialized before the calls to `DebugLogger` at lines 68 and 77.

(Which, when I tried to incorporate the resulting GDNative library into my FO Godot prototype, caused my prototype to crash immediately without any reports about what went wrong. Turned out the problem was that I didn't provide a proper path to the `default` folder which threw the runtime error, but I got no log which would have told me what was the problem.)

Reply:
I got logs in console when miss `default` directory. But I'll move up them.

Reply:
Unfortunately I didn't. Whatever gets printed to `stdout` or `stderr` seems to get lost. Or maybe I don't know where to look in the Godot IDE... 😕

Reply:
Also the original client doesn't initialize application before directory check. I suppose something else was missing.

Reply:
I launched game directly not from IDE so I got its `stdout` and `stderr`.

Reply:
Yes, IDE doesn't capture `stdout` and `stderr` of debugging process.

Reply:
Moved app initialization up.